cpp/src/arrow/array/array_encoded.cc

```diff
@@ -0,0 +1,79 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/array/array_encoded.h"
+#include "arrow/array/util.h"
+#include "arrow/util/logging.h"
+#include "arrow/util/ree_util.h"
+
+namespace arrow {
+
+// ----------------------------------------------------------------------
+// RunEndEncodedArray
+
+RunEndEncodedArray::RunEndEncodedArray(const std::shared_ptr<ArrayData>& data) {
+  ARROW_CHECK_EQ(data->type->id(), Type::RUN_END_ENCODED);
+  SetData(data);
+}
+
+RunEndEncodedArray::RunEndEncodedArray(const std::shared_ptr<DataType>& type,
+                                       int64_t logical_length,
+                                       const std::shared_ptr<Array>& run_ends,
+                                       const std::shared_ptr<Array>& values,
+                                       int64_t offset) {
+  ARROW_CHECK_EQ(type->id(), Type::RUN_END_ENCODED);
+  SetData(ArrayData::Make(type, logical_length,
+                          /*buffers=*/{NULLPTR},
+                          /*child_data=*/{run_ends->data(), values->data()}, 0, offset));
+}
+
+Result<std::shared_ptr<RunEndEncodedArray>> RunEndEncodedArray::Make(
+    const std::shared_ptr<Array>& run_ends, const std::shared_ptr<Array>& values,
+    int64_t logical_length, int64_t offset) {
+  auto run_ends_type = run_ends->type();
+  auto values_type = values->type();
+  if (!RunEndEncodedType::RunEndsTypeValid(*run_ends_type)) {
+    return Status::Invalid("Run ends array must be int16, int32 or int64 type");
+  }
+  if (run_ends->null_count() != 0) {
+    return Status::Invalid("Run ends array cannot contain null values");
+  }
+
+  return std::make_shared<RunEndEncodedArray>(
+      run_end_encoded(std::move(run_ends_type), std::move(values_type)), logical_length,
+      run_ends, values, offset);
+}
+
+std::shared_ptr<Array> RunEndEncodedArray::run_ends_array() const {
+  return MakeArray(data()->child_data[0]);
+}
+
+std::shared_ptr<Array> RunEndEncodedArray::values_array() const {
+  return MakeArray(data()->child_data[1]);
+}
```

Comment:
These could be cached on the instance instead of calling `MakeArray` every time.